sync.py

```diff
@@ -30,13 +35,183 @@ def __init__(self):
         self.codes_found_in_cascade = []
         self.missing_data_codes = []
 
-    def process_all_blocks(self, wsapi_data, time_to_wait, send_email_after):
+    def convert_dictionary_to_hash(self, dictionary):
+
+        def _recursively_alphabetize_dictionary_by_keys(dictionary_to_alphabetize):
+            if isinstance(dictionary_to_alphabetize, dict):
+                to_return = '{'
+                ordered_keys = sorted(dictionary_to_alphabetize.keys())
+                for key in ordered_keys:
+                    to_return += "'%s': " % str(key)
+                    value = dictionary_to_alphabetize[key]
+                    if isinstance(value, dict):
+                        to_return += _recursively_alphabetize_dictionary_by_keys(value)
+                    elif isinstance(value, (str, unicode)):
+                        to_return += "'%s'" % str(value)
+                    to_return += ', '
+
+                return to_return + '}'
+            else:
+                return 'Not a dictionary'
+
+        alphabetized_string = _recursively_alphabetize_dictionary_by_keys(dictionary)
+
+        return repr(hashlib.md5(alphabetized_string).digest())
+
+    def get_changed_banner_rows_and_distinct_prog_codes(self):
+        """
+        In order for this method to work properly, the rows of data returned by Banner via WSAPI must be sorted
+        alphabetically by program code. If the rows are parsed out of order, the tertiary check (row index) will
+        produce errant results. Thankfully the sorting is already done by WSAPI before the data gets sent here.
+        :return (Dict[] changed_rows, Set distinct_prog_codes)
+        """
+
+        # First, read in dictionary of old {row_keys: md5 hashes} from .csv
+        # Disclaimer: md5 hashes can contain commas, so the separator is technically ',\t'
+        old_hashes = {}
+        if os.path.isfile(BANNER_HASHES_CSV_PATH):
+            with open(BANNER_HASHES_CSV_PATH, 'r') as old_data_hashes:
+                for line in old_data_hashes.readlines():
+                    vals = line.split(',\t')
+                    old_hashes[vals[0]] = vals[1]
+
+        # Second, read in current values of data from Banner via WSAPI
+        # These are sorted twice in WSAPI: first by prog_code, and then sub-sorted by start_term_code
+        new_banner_data = json.loads(requests.get('https://wsapi.bethel.edu/program-data').content)
+        # Sourced from: https://stackoverflow.com/a/16839304
+        data_len_magnitude = int(math.log10(len(new_banner_data)))
```

Comment:
I think you are overcomplicating the prepending zeros by using log. Try something like this: https://stackoverflow.com/questions/339007/nicest-way-to-pad-zeroes-to-a-string

Reply:
Good catch. Changed.